lib/remote_node.go

```diff
@@ -0,0 +1,744 @@
+	vMeta := rn.handshakeMetadata
+
+	// Verify that the verack message is formatted correctly according to the PoS standard.
+	if vrkMsg.Version != VerackVersion1 {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack version mismatch; message: %v; expected: %v", rn.id, vrkMsg.Version, VerackVersion1)
+	}
+
+	// Verify that the counterparty's verack message's NonceReceived matches the NonceSent we sent.
+	if vrkMsg.NonceReceived != vMeta.versionNonceSent {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) nonce mismatch; "+
+			"message: %v; nonceSent: %v", rn.id, vrkMsg.NonceReceived, vMeta.versionNonceSent)
+	}
+
+	// Verify that the counterparty's verack message's NonceSent matches the NonceReceived we sent.
+	if vrkMsg.NonceSent != vMeta.versionNonceReceived {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack nonce mismatch; message: %v; expected: %v", rn.id, vrkMsg.NonceSent, vMeta.versionNonceReceived)
+	}
+
+	// Get the current time in microseconds and make sure the verack message's timestamp is within 15 minutes of it.
+	timeNowMicro := uint64(time.Now().UnixMicro())
+	if vrkMsg.TstampMicro < timeNowMicro-rn.params.HandshakeTimeoutMicroSeconds {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack timestamp too far in the past. Time now: %v, verack timestamp: %v", rn.id, timeNowMicro, vrkMsg.TstampMicro)
+	}
+
+	// If the RemoteNode is not a validator, then we don't need to verify the verack message's signature.
+	if !rn.hasValidatorServiceFlag() {
+		return nil
+	}
+
+	// Make sure the verack message's public key and signature are not nil.
+	if vrkMsg.PublicKey == nil || vrkMsg.Signature == nil {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack public key or signature is nil", rn.id)
+	}
+
+	// Verify the verack message's signature.
+	ok, err := BLSVerifyPoSValidatorHandshake(vrkMsg.NonceSent, vrkMsg.NonceReceived, vrkMsg.TstampMicro,
+		vrkMsg.Signature, vrkMsg.PublicKey)
+	if err != nil {
+		return errors.Wrapf(err, "RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack signature verification failed with error", rn.id)
+	}
+	if !ok {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack signature verification failed", rn.id)
+	}
+
+	if rn.validatorPublicKey != nil && rn.validatorPublicKey.Serialize() != vrkMsg.PublicKey.Serialize() {
+		return fmt.Errorf("RemoteNode.validateVerackPoS: Requesting disconnect for id: (%v) "+
+			"verack public key mismatch; message: %v; expected: %v", rn.id, vrkMsg.PublicKey, rn.validatorPublicKey)
+	}
+
+	// If we get here then the verack message is valid. Set the validator public key on the peer.
+	vMeta.validatorPublicKey = vrkMsg.PublicKey
```

Comment:
This is a weird thing to do. Didn't you already check that rn.validatorPublicKey == vrkMsg.PublicKey explicitly above?